### PR TITLE
Move logic from hardhat patches

### PR DESCRIPTION
**Pull request by yivlad**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -36,31 +36,30 @@
...
     "ts-node": "^8.1.0",
-    "typescript": "~4.5.2"
+    "typescript": "~4.5.2",
+    "@types/sinon-chai": "^3.2.3",
+    "@types/web3": "1.0.19"
   },
   "peerDependencies": {
     "@nomiclabs/hardhat-ethers": "^2.0.0",
-    "ethereum-waffle": "^3.2.0",
+    "ethereum-waffle": "*",
```
Comment (rzadp) [inline package.json:63]:
```suggestion
    "ethereum-waffle": ">=4.0.7",
```
How about this?

File: test/fixture-projects/hardhat-project/test/tests.js [hunk 2]
```diff
@@ -94,8 +68,8 @@ describe("Internal test suite of hardhat-waffle's test project", function () {
       await contract.deployed();
 
       await expect(() =>
-        contract.incByValue({ value: 200 })
-      ).to.changeEtherBalance(sender, -56189212266592, { includeFee: true });
+        contract.incByValue({ value: 200, gasPrice: 1265506330 })
```
Comment (rzadp) [inline test/fixture-projects/hardhat-project/test/tests.js]: Is the gasPrice setting here necessary?

File: test/fixture-projects/configs/shared-tests/skip-estimate-gas.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      });
+      const gas2 = await waffle.provider.estimateGas({
+        to: contract.address,
+        data: calldata2,
+      });
+
+      if (enabled) {
+        expect(
+          gas1.toHexString() === gas2.toHexString() &&
+            gas1.toHexString() === "0xB71B00",
```
Comment (rzadp) [inline test/fixture-projects/configs/shared-tests/skip-estimate-gas.js]:
```suggestion
            gas1.toHexString() === "0xB71B00", // This value is set in `config.waffle.skipEstimategas`.
```

File: package.json (hunk 1 above)
Comment (yivlad) [reply package.json:63]: I think the plugin will also work with waffle 3.0, why putting additional constraints?

File: test/fixture-projects/hardhat-project/test/tests.js (hunk 2 above)
Comment (yivlad) [reply test/fixture-projects/hardhat-project/test/tests.js]: I don't know why but hardhat seems to change gasPrice depending on the state of the blockchain. This leads to the test being unpredictable if gasPrice is not fixed

File: package.json (hunk 1 above)
Comment (rzadp) [reply package.json:63]: right, makes sense.

File: test/fixture-projects/hardhat-project/test/tests.js (hunk 2 above)
Comment (rzadp) [inline test/fixture-projects/hardhat-project/test/tests.js]:
```suggestion
        contract.incByValue({ value: 200, gasPrice: 1265506330 }) // Constant gasPrice in order to make tests predictable.
```

**Review by rzadp — APPROVED** [review]

**Comment by corygabrielsen** [comment]
hey there guys @yivlad @rzadp 
I think this commit introduced a packaging bug (or I'm dumb?) because I cannot upgrade from `2.0.1` without adding `"@types/sinon-chai": ""^3.2.3"` to my devDependencies explicitly.

I spent about an hour debugging it and whenever I upgrade the package, I need to explicitly re-add the devDependencies on `@types/sinon-chai`

Here is the build error I get without the fix:
```
> @redacted/packagename
$ yarn hardhat compile
$ /path/to/node_modules/.bin/hardhat compile
Compiled 100 Solidity files successfully
$ /path/to/node_modules/.bin/tsc
../node_modules/@ethereum-waffle/chai/dist/esm/types.d.ts:2:23 - error TS2688: Cannot find type definition file for 'sinon-chai'.

2 /// <reference types="sinon-chai" />
                        ~~~~~~~~~~


Found 1 error in ../node_modules/@ethereum-waffle/chai/dist/esm/types.d.ts:2

error Command failed with exit code 2.
```

So in order to build your package it is looking for those types. and users have to figure out they need the dependency in dev dependencies.

**Comment by yivlad** [comment]
Hey @stoooops,

what version of `ethereum-waffle` are you using? I'm wondering cause we no longer have dependency to sinon-chai specified in that way

**Pull request merged**